Deprecate `Currency`; introduce holds and freezing into `fungible` traits

gavofyork opened the pull request:
Related to https://github.com/paritytech/polkadot-sdk/issues/236
Fixes https://github.com/paritytech/substrate/issues/12701
Polkadot Companion: https://github.com/paritytech/polkadot/pull/6780
Cumulus Companion: https://github.com/paritytech/cumulus/pull/2334

Introduces Freezing and Holds: two new facilities in Balances pallet and an interface for them through `fungible` and `fungibles`. Also reworks some internals of Balances pallet to make it safer. Requires Existential Deposit requirements to be made of the free balance, not the total balance.

## Key Points

- Introduces 4 new items in `pallet_balances::Config` trait:
  - `type MaxHolds`
  - `type MaxFreezes`
  - `type HoldIdentifier`
  - `type FreezeIdentifier`
  - Until you introduce pallets which use holds and freezes, then these can safely be set to `ConstU32<0>` in the case of the first two and `()` in the case of the second two.
- Introduces a new dispatchable (extrinsic) in `pallet_balances`: `upgrade_accounts`.
- Renames 2 dispatchables:
  - `Balances::set_balance` becomes `Balances::force_set_balance` and only accepts the free balance (not the reserved balance since this no longer makes sense).
  - `Balances::transfer` becomes `Balances::transfer_allow_death`
  - Uses of both within your codebases should be renamed as you upgrade the version of your Balances pallet.
- Alters many function signatures of `fungible`/`fungibles` traits making them much more type-safe.
- Merges the `fungible::Transfer` trait into `fungible::Mutate` (and same for `fungibles`).
- Requires the `fungible::Balanced` trait to be explicitly implemented (and same for `fungibles`).
- Introduces default implementations for almost all functionality in the `fungible`/`fungibles` traits.
- Silently deprecates `Currency` traits in favour of `fungible` traits.
- Accounts now require the Existential Deposit (`Currency::minimum_balance()`) **in addition to** any amount to be reserved. This is a **BREAKING CHANGE** and is likely to result in failures in tests and benchmarking.
- An `ExistentialDeposit` of zero no longer works. If you need to allow any account to have a zero balance and retain associated account data, you must introduce a pallet which allows an unpermissioned `frame_system::Pallet::<Runtime>::inc_providers()`.

## Long Description

Some renaming:

- `Balances::set_balance` becomes `Balances::force_set_balance`
- `Balances::transfer` becomes `Balances::transfer_allow_death`

In the latter case, since ecosystem tooling largely depends on using `transfer`, a compatibility stub for `transfer` remains, however **this naming is DEPRECATED** and ecosystem tools should move over to using `transfer_allow_death` **within 3 months** at which point the `transfer` alias will be removed.

In general, it's now better to use the `fungible` trait API for Balances rather than messing with the dispatchables directly. The long-term plan is to remove the dispatchables and have all pallet-access be through XCM.

### Holds and Freezes

Holds are roughly analagous to `reserve`s, but are now explicitly designed to be infallibly slashed. They do not contribute to the ED but do require a provider reference, removing any possibility of account reference counting from being problematic for a `slash`. They are also always named, ensuring different holds do not accidentally slash each other's balances.

Freezes are essentially the same as locks, except that they overlap with Holds. Since Holds are designed to be infallibly slashed, this means that any logic using a Freeze must handle the possibility of the frozen amount being reduced, potentially to zero. A permissionless function should be provided in order to allow bookkeeping to be updated in this instance.

Both Holds and Freezes require an identifier which is configurable and is expected to be an `enum` aggregated across all pallet instances of the runtime.

### Balances Internals

Alter `reserve` functionality in Balances so that it *does not* contribute to the account's existential deposit and *does* use a consumer reference. Remove the concept of withdraw reasons from locks. Migration is lazy: accounts about to be mutated will be migrated prior. New accounts will automatically be mutated. A permissionless upgrade dispatchable is provided to upgrade dormant accounts.

### Typing

Type-safety is improved by removing all naked `bool` parameters from fungible(s) API functions. In particular, `best_effort` and `force` which were both previously boolean are now typed as `Precision` and `Fortitude`. Lesser used arguments which have changed are `on_hold` which is typed as `Restriction` and `minted` which is typed as `Provenance`. This generally makes invocations much more readable.

### Privatizations

One function `mutate_account` (in Balances pallet) has been privatized. It should never have been public since it ignores some important bookkeeping. If you're using it, you probably mean to be using `make_free_balance_be` instead.

## Upgrading

Because reserved/held funds do not any longer contribute to the ED, you may find that certain test and benchmark code no longer works. This is generally due to the case where exactly as many funds were placed in an account as were later reserved. To fix this breakage, simply ensure that at least the currency's `minimum_deposit` is in the account in addition to any funds which are planned to be reserved. It's usually as simple as appending `+ T::Currency::minimum_balance()` to the argument of `make_free_balance_be`.

In some circumstances, errors have changed to the more general errors.

After introducing a runtime with this change in it, chains should ensure that all accounts are upgraded by repeatedly calling `update_accounts` giving it account IDs which are not yet upgraded. A script which does this shall be provided. 

Pallets which use the existing set of `Currency` traits should be refactored to use the `fungible` traits. Aside from the actual code change, a gateway function (`ensure_upgraded`) should be introduced and called prior to any mutating operations to ensure that any accounts using the old `Currency` traits are unreserved/unlocked and said funds are instead placed on hold/frozen. A permissionless fee-free dispatchable should also be introduced (`upgrade_accounts`), similar to that of the Balances pallet, which is able to do this for dormant accounts.

It is not crucial to get this done in any particular timeframe, however at some point in the future Currency traits, Locks and Reserves will be deprecated and removed, so it will need to happen before then.

### Important Notes

Some functions in the `Currency` traits retain their name and parameter set in the `fungible` traits and have similar semantics. However, there are many functions with differences in naming, and some with differences in result types and their meaning.

- `Currency::free_balance` may be replaced with `fungible::Inspect::balance`.
- `Currency::make_free_balance_be` may be replaced with `fungible::Mutate::set_balance`.
- `ReservableCurrency::reserved_balance` may be replaced with `fungible::hold::Inspect::total_balance_on_hold`.
- `NamedReservableCurrency::reserve` may be replaced with `fungible::hold::Mutate::hold`.
- `NamedReservableCurrency::unreserve` may be replaced with `fungible::hold::Mutate::release`, **however** the returned `Result` must be handled and the inner of `Ok` has the opposite meaning (i.e. it is the amount released, rather than the amount *not* released, if any).
- `LockableCurrency::set_lock` may generally be replaced with `fungible::freeze::Mutate::set_freeze`.
- `LockableCurrency::extend_lock` may generally be replaced with `fungible::freeze::Mutate::extend_freeze`.
- `LockableCurrency::remove_lock` may be replaced with `fungible::freeze::Mutate::thaw`.

There are no equivalents for `ReservableCurrency::reserve` or `::unreserve`: an identifier must always be used.

Functions which may result in an account being deleted accept a new `Preservation` value. This has three variants:

- `Expendable`: Equivalent to `ExistenceRequirement::AllowDeath`
- `Protect`: The account must not die. If there are multiple provider refs, then this may still allow the balance to be reduced to zero.
- `Preserve`: Not only must the account not die, but the provider reference of the Currency implementation must also be kept in place.

### Pallet NIS

NOTE: This updates NIS to use only the `fungible` API rather than `Currency`. This is fine since NIS is not deployed anywhere. For other pallets already in production then a progressive migration will be needed.

**Pallet NIS has been altered to use `active_issuance` rather than `total_issuance`. This means the `IgnoredIssuance` is whatever should be discounted against `active_issuance`. _It should probably be replaced with `()`._**

## TODO

- [x] Create `UnbalancedHold`.
- [x] Implement `UnbalancedHold` for balances.
- [x] Implement all the other `Hold` traits for `UnbalancedHold` implementations.
- [x] Deprecate `fee_frozen` and `WithdrawReasons`; just have a single `frozen` field in `AccountData`.
- [x] Impl `InspectFreeze` and `MutateFreeze` for Balances.
  - [x] Main implementation
  - [x] Tests for `MutateFreeze`
  - [x] Tests for combination of freezing and locking
- [x] Support arbitrary IDs for freezes and holds.
- [x] Change freeze/hold mechanism to not be mutually exclusive.
  - [x] Introduce tests for this.
- [x] Permissionless fee-free dispatchable for running `ensure_upgraded` on a non-upgraded account.
  - [x] Benchmarking
  - [x] Test
- [x] Refactor `fungible`/`fungibles`
  - [x] Remove all blanket impls.
  - [x] Events for all mutator traits.
  - [x] Bring `fungibles` up to date.
  - [x] Make `ItemOf` work properly again.
  - [x] Make Assets pallet build.
- [x] Refactor `fungible`/`fungibles`/`nonfungibles`:
  - [x] Rename `suspend`/`resume` to `shelve`/`restore`.
- [x] Fix tests using Balances pallet owing to reserves now requiring a consumer ref and not counting toward ED.
- [x] Script for calling `upgrade_accounts`: https://github.com/ggwpez/substrate-scripts/blob/master/upgrade-accounts.py
- [x] Handle dust in legacy uses of mutate account.
- [x] Fix up remaining usages of old bool API.
- [x] Fix up assets and balances benchmarks.



## Maybe
- [ ] Make `Currency` an optional feature.

Next up: https://github.com/paritytech/polkadot-sdk/issues/226

gavofyork committed:
First reworking of fungibles API

gavofyork committed:
New API and docs

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2166714

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2166724

gavofyork committed:
More fungible::* API improvements

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2169176

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2169175

gavofyork committed:
New ref-counting logic for old API

gavofyork committed:
Missing files

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2174970

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2174975

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2174974

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2174969

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2175681

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2175685

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2175686

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2175680

gavofyork committed:
Use the new transfer logic

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2179352

gavofyork committed:
Use fungibles for the dispatchables

gavofyork committed:
Use shelve/restore names

gavofyork committed:
Locking works with total balance.

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2194365

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2194366

gavofyork committed:
repotting and removal

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2202930

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2202931

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2202945

gavofyork committed:
Separate Holds from Reserves

gavofyork committed:
Introduce freezes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2274746

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
Missing files

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2283677

gavofyork committed:
Tests for freezing

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2284083

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2284084

gavofyork committed:
Fix hold+freeze combo

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2284795

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2284794

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2284809

gavofyork committed:
More tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2288898

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2288899

gavofyork committed:
Fee-free dispatchable for upgrading accounts

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2288997

gavofyork committed:
Benchmarks and a few fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2290146

gavofyork committed:
Another test

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2292172

gavofyork committed:
Docs and refactor to avoid blanket impls

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2292810

gavofyork committed:
Repot

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2292912

gavofyork committed:
Fit out ItemOf fully

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2294255

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2294254

gavofyork committed:
Add events to Balanced traits

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2294303

gavofyork committed:
Introduced events into Hold traits

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2294352

gavofyork committed:
Fix Assets pallet tests

gavofyork committed:
Assets benchmarks pass

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2294631

gavofyork committed:
Missing files and fixes

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2297068

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2300341

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2300343

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2300342

gavofyork committed:
Benchmarks fixes

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
Fix balance benchmarks

gavofyork committed:
Move NIS to fungible API

gavofyork committed:
Formatting

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2304162

ggwpez commented on frame/balances/src/lib.rs:
```diff
@@ -451,300 +664,69 @@ pub mod pallet {
+			}
+			let mut upgrade_count = 0;
+			for i in &who {
+				let upgraded = Self::ensure_upgraded(i);
```
Some `Upgraded` event here or in the `ensure_upgraded` function would be nice.

ggwpez commented on frame/balances/src/lib.rs:
```diff
@@ -257,6 +276,217 @@ pub mod pallet {
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config<I>, I: 'static = ()> {
+		pub balances: Vec<(T::AccountId, T::Balance)>,
```
It looks like you are setting the frozen/reserved balance in tests manually, which is fine.  
But for tests on a test-net it is nice to have the frozen/reserved also in the genesis spec.

gavofyork replied:
This is not supported.

gavofyork committed:
Expose fungible sub modules

gavofyork committed:
Merge branch 'gav-hold-freeze' into gav-nis-hold

gavofyork committed:
Fix broken impl and add test

gavofyork committed:
Merge branch 'gav-hold-freeze' into gav-nis-hold

gavofyork committed:
Fix tests

gavofyork committed:
Use composite APIs

gavofyork committed:
API for `transfer_and_hold`

gavofyork committed:
Merge branch 'gav-hold-freeze' into gav-nis-hold

gavofyork committed:
Merge branch 'gav-nis-hold' into gav-hold-freeze

gavofyork committed:
Formatting

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2305302

gavofyork committed:
Upgraded event

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2305395

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2305590

gavofyork committed:
Fixes

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2305932

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2306172

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2306173

gilescope commented:
Would this be a good opportunity to rename the Mutate trait from `make_balance_be` to `set_balance` to be more consistent and friendly to newcomers?

gavofyork committed:
Repot tests and some fixed

gavofyork committed:
Fix some bits

gavofyork committed:
Fix dust tests

gavofyork committed:
Rename `set_balance`

- `Balances::set_balance` becomes `Balances::force_set_balance`
- `Unbalanced::set_balance` becomes `Unbalances::write_balance`

gavofyork committed:
becomes

gavofyork committed:
Move dust handling to fungibles API

gavofyork committed:
Formatting

gavofyork committed:
Merge branch 'master' into gav-hold-freeze

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2325555

gavofyork committed:
Fixes and more refactoring

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2326102

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2326800

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2327008

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2327260

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2327259

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2328035

gavofyork committed:
Fixes

gavofyork committed:
Merge branch 'master' into gav-hold-freeze

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2328314

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2328315

gavofyork committed:
Fixes

gavofyork committed:
Merge branch 'gav-hold-freeze' of github.com:paritytech/substrate into gav-hold-freeze

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2329053

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2329052

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2329106

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2329170

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2329172

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2329171

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2329237

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2329238

muharem commented on frame/support/src/traits/tokens/fungible/hold.rs:
```diff
@@ -0,0 +1,385 @@
+
+	/// Get the maximum amount that the `total_balance_on_hold` of `who` can be reduced successfully
+	/// based on whether we are willing to force the reduction and potentially go below user-level
+	/// restrictions on the minimum amount of the account.
```
where `minimum_balance` has no effect, only frozen (according to the impl)

muharem commented on frame/balances/src/lib.rs:
```diff
@@ -608,1602 +466,615 @@ pub mod pallet {
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let who = T::Lookup::lookup(who)?;
+			let _leftover = <Self as ReservableCurrency<_>>::unreserve(&who, amount);
```
Could not this lead to some inconsistency? Since no reason given,  no attempt to update the `Holds`.
Still need to understand the migration process better.

gavofyork replied:
Yes - this is really for pre-existing anonymous reserves under the old API. It will be removed once all pallets have been transitioned to the new API.

muharem reviewed (commented):
From the person who has a little experience in the domain, the new naming and model, new API and its implementation, looks correct to me and as a big improvement.

gavofyork committed:
Use reducible_balance for better correctness on fees

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2360259

gavofyork committed:
Reducing hold to zero should remove entry.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2360343

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2360344

gavofyork committed:
Add test

gavofyork committed:
Docs

gavofyork committed:
Update frame/support/src/traits/tokens/fungibles/hold.rs

Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungibles/regular.rs

Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungible/hold.rs

Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungible/regular.rs

Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

gavofyork committed:
Docs

gavofyork committed:
Merge branch 'gav-hold-freeze' of github.com:paritytech/substrate into gav-hold-freeze

gavofyork committed:
Docs

gavofyork committed:
Docs

gavofyork commented:
@muharem All points addressed - thanks for the review.

gavofyork committed:
Merge branch 'master' into gav-hold-freeze

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2362188

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2362189

muharem commented on frame/balances/src/lib.rs:
```diff
@@ -608,1602 +466,611 @@ pub mod pallet {
-				amount = *issued;
-				Zero::zero()
+				let maybe_endowed = if is_new { Some(account.free) } else { None };
+				let maybe_account_maybe_dust = Self::post_mutation(who, account);
```
Might be simpler if we drop `post_mutation` abstraction. 
It's not reused, and visible only within the crate.

gavofyork committed:
Fix NIS benchmarks

gavofyork commented on frame/assets/src/functions.rs:
```diff
@@ -76,7 +76,12 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
-			frame_system::Pallet::<T>::inc_consumers(who).map_err(|_| Error::<T, I>::NoProvider)?;
+			frame_system::Pallet::<T>::inc_consumers(who)
+				.map_err(|_| Error::<T, I>::UnavailableConsumer)?;
+			ensure!(
```
No - I added a comment to explain.

gavofyork committed:
Doc comment

gavofyork commented on frame/balances/src/lib.rs:
```diff
@@ -608,1602 +466,611 @@ pub mod pallet {
+				if !did_provide && does_provide {
+					frame_system::Pallet::<T>::inc_providers(who);
+				}
+				if did_consume && !does_consume {
+					frame_system::Pallet::<T>::dec_consumers(who);
+				}
+				if !did_consume && does_consume {
+					frame_system::Pallet::<T>::inc_consumers(who)?;
+				}
```
Having funds on hold/reserved or having freezes/locks active requires a consumer reference since it uses additional storage in the Balances pallet.

Staking also uses up storage so it requires a consumer reference. As it happens, staking also uses locks (soon to be updated to a hold), so it ends up using two references. But the reason for those two references are both valid and independent.

muharem replied:
Understood!

gavofyork committed:
Remove post_mutation

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2370116

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2370117

gavofyork commented:
@muharem please confirm you're happy with the final queries.

ruseinov commented on frame/vesting/src/tests.rs:
```diff
@@ -180,10 +183,11 @@ fn unvested_balance_should_not_transfer() {
-			pallet_balances::Error::<Test, _>::LiquidityRestrictions,
-		); // Account 1 cannot send more than vested amount
+		assert_noop!(Balances::transfer_allow_death(Some(1).into(), 2, 56), TokenError::Frozen,); // Account
+		                                                                                  // 1 cannot
```
seems like there is a new line issue with this comment.

ruseinov commented on primitives/runtime/src/lib.rs:
```diff
@@ -606,9 +606,9 @@ impl From<crate::traits::BadOrigin> for DispatchError {
-	/// Account that must exist would die.
-	WouldDie,
+	FundsUnavailable,
+	/// Balance is needed to fund a needed provider reference.
```
This comment is a bit cryptic, at least to me. So is the error name. Perhaps rephrasing this could make things easier to understand. 
Does this mean that the leftover balance is needed to fund the provider?

ruseinov commented on frame/transaction-payment/asset-tx-payment/src/payment.rs:
```diff
@@ -126,8 +126,15 @@ where
 		}
-		<T::Fungibles as Balanced<T::AccountId>>::withdraw(asset_id, who, converted_fee)
-			.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))
+		<T::Fungibles as Balanced<T::AccountId>>::withdraw(
```
just an observation: the list of `withdraw` params has grown to a point that it is hard to follow, at least the boolean flags.

gavofyork replied:
Fixed.

ruseinov commented on frame/system/src/lib.rs:
```diff
@@ -1700,23 +1696,12 @@ impl<T: Config> StoredMap<T::AccountId, T::AccountData> for Pallet<T> {
 		let account = Account::<T>::get(k);
-		let was_providing = is_providing(&account.data);
-		let mut some_data = if was_providing { Some(account.data) } else { None };
+		let was_something = account.data != T::AccountData::default();
+		let mut some_data = if was_something { Some(account.data) } else { None };
```
```suggestion
		let is_empty = account.data == T::AccountData::default();
		let mut some_data = if !is_empty { Some(account.data) } else { None };
```

vivekvpandya commented on frame/assets/src/lib.rs:
```diff
@@ -773,6 +773,7 @@ pub mod pallet {
 			let origin = ensure_signed(origin)?;
 			let beneficiary = T::Lookup::lookup(beneficiary)?;
 			let id: T::AssetId = id.into();
+
```
May be unintended new line?

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
Fix some tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2435062

gavofyork committed:
Fix some grumbles

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2437020

gavofyork committed:
Enumify bool args to fungible(s) functions

gavofyork committed:
Fix up assets and balances

gavofyork committed:
Formatting

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2437557

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2437556

gavofyork committed:
Fix contracts

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2438379

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
Fix tests & benchmarks build

gavofyork committed:
Typify minted boolean arg

gavofyork committed:
Typify on_hold boolean arg; renames

gavofyork committed:
Fix numerous tests

gavofyork committed:
Fix dependency issue

gavofyork committed:
Privatize dangerous API mutate_account

athei committed:
Fix tests for contracts

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2440892

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2440888

gavofyork committed:
Fix contracts (@alext - please check this commit)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2441173

gavofyork commented:
@ruseinov please check changes are satisfactory and approve if so.

gavofyork commented on frame/contracts/src/tests.rs:
```diff
@@ -1573,7 +1577,7 @@ fn call_return_code() {
 		// threshold when transfering 100 balance but this balance is reserved so
 		// the transfer still fails.
 		Balances::make_free_balance_be(&addr_bob, min_balance + 100);
-		Balances::reserve(&addr_bob, min_balance + 100).unwrap();
+		Balances::reserve(&addr_bob, 100).unwrap();
```
Not sure which part of the test you mean (I'm not very familiar with the code) - could you elaborate?

athei replied:
Pushed the change.

gavofyork committed:
Remove println

gavofyork commented:
@kianenigma could do with a review.

gavofyork committed:
Fix broken rename

gavofyork committed:
Merge branch 'gav-hold-freeze' of github.com:paritytech/substrate into gav-hold-freeze

gavofyork committed:
Fix broken rename

gavofyork committed:
Fix broken rename

gavofyork committed:
Docs

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2442467

jsidorenko commented on frame/assets/src/impl_fungibles.rs:
```diff
@@ -65,69 +77,24 @@ impl<T: Config<I>, I: 'static> fungibles::Inspect<<T as SystemConfig>::AccountId
 	for Pallet<T, I>
 {
-	/// Return the name of an asset.
-	fn name(asset: &Self::AssetId) -> Vec<u8> {
```
What's the reason for removing those methods? This could easily break existing integrations, not to say now there are no public methods to get that info on the pallet level (e.g. another pallet would need to create an asset and then read/validate the name/symbol/decimals)

gavofyork replied:
Metadata can be inspected and mutated through the traits in the metadata module. I apparently never added the file - will do so now.

jsidorenko replied:
Sorry @gavofyork , what `metadata` module you're referencing to? If it's the Metadata storage item, then it's not public and thus can't be accessed from another pallet.

tonyalaribe commented on frame/support/src/traits/tokens/fungible/hold.rs:
```diff
@@ -0,0 +1,393 @@
+		// check liquidity.
+
+		Self::ensure_can_hold(reason, who, amount)?;
+		// Should be infallible now, but we proceed softly anyway.
```
This comment is a bit unclear. `We proceed softly` but the Fortitude is `Force` which doesn't sound soft.

tonyalaribe commented on frame/support/src/traits/tokens/fungible/hold.rs:
```diff
@@ -0,0 +1,393 @@
+	/// Check to see if some `amount` of funds of `who` may be placed on hold with the given
+	/// `reason`. Reasons why this may not be true:
+	///
+	/// - The implementor supports only a limited number of concurrernt holds on an account which is
```
```suggestion
	/// - The implementor supports only a limited number of concurrent holds on an account which is
```

gavofyork committed:
Update frame/support/src/traits/tokens/fungible/hold.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2479015

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
remove from_ref_time

tonyalaribe commented:
A tiny update to the description for this PR

The line:
> Functions which may result in an account being deleted accept a new **KeepAlive** value. This has three variants:


Should be 
> Functions which may result in an account being deleted accept a new **Preservation** value. This has three variants:

since it's no longer KeepAlive, but `Preservation` in the latest changes.

tonyalaribe commented on frame/support/src/traits/tokens/fungibles/hold.rs:
```diff
@@ -0,0 +1,455 @@
+	/// Check to see if some `amount` of funds of `who` may be placed on hold with the given
+	/// `reason`. Reasons why this may not be true:
+	///
+	/// - The implementor supports only a limited number of concurrernt holds on an account which is
```
```suggestion
	/// - The implementor supports only a limited number of concurrent holds on an account which is
```

tonyalaribe commented on frame/support/src/traits/tokens/fungibles/hold.rs:
```diff
@@ -0,0 +1,455 @@
+	/// - Removing `amount` from the total balance would kill the account and remove the only
+	///   provider reference.
+	///
+	/// Note: we pass `true` as the third argument to `reducible_balance` since we assume that if
+	/// needed the balance can slashed. If we are using a simple non-forcing reserve-transfer, then
+	/// we really ought to check that we are not reducing the funds below the freeze-limit (if any).
```
```suggestion
	/// Note: we pass `Fortitude::Force` as the last argument to `reducible_balance` since we assume that if
	/// needed the balance can slashed. If we are using a simple non-forcing reserve-transfer, then
	/// we really ought to check that we are not reducing the funds below the freeze-limit (if any).
```

tonyalaribe commented on frame/support/src/traits/tokens/fungibles/hold.rs:
```diff
@@ -0,0 +1,455 @@
+	/// Check to see if some `amount` of funds of `who` may be placed on hold for the given
+	/// `reason`. Reasons why this may not be true:
+	///
+	/// - The implementor supports only a limited number of concurrernt holds on an account which is
```
```suggestion
	/// - The implementor supports only a limited number of concurrent holds on an account which is
```

tonyalaribe commented on frame/support/src/traits/tokens/fungibles/hold.rs:
```diff
@@ -0,0 +1,455 @@
+
+	/// Reduce the balance on hold of `who` by `amount`.
+	///
+	/// If `precision` is `false` and it cannot be reduced by that amount for
+	/// some reason, return `Err` and don't reduce it at all. If `precision` is `BestEffort`, then
+	/// reduce the balance of `who` by the most that is possible, up to `amount`.
```
```suggestion
	/// If `precision` is `Precision::Exact` and it cannot be reduced by that amount for
	/// some reason, return `Err` and don't reduce it at all. If `precision` is `Precision::BestEffort`, then
	/// reduce the balance of `who` by the most that is possible, up to `amount`.
```

tonyalaribe commented on frame/support/src/traits/tokens/currency.rs:
```diff
@@ -176,7 +176,7 @@ pub trait Currency<AccountId> {
 
 	/// Removes some free balance from `who` account for `reason` if possible. If `liveness` is
-	/// `KeepAlive`, then no less than `ExistentialDeposit` must be left remaining.
+	/// `Expendability`, then no less than `ExistentialDeposit` must be left remaining.
```
it says:
> if `liveness` is `Expendability`
but `liveness` (`ExistenceRequirement`) can only be either `KeepAlive` or `AllowDeath`
```suggestion
	/// `KeepAlive`, then no less than `ExistentialDeposit` must be left remaining.
```

tonyalaribe commented on frame/support/src/traits/tokens/fungible/hold.rs:
```diff
@@ -0,0 +1,393 @@
+	/// If `precision` is `BestEffort`, then an amount less than `amount` may be transferred without
+	/// error.
+	///
+	/// If `force` is `Force`, then other fund-locking mechanisms may be disregarded. It should be
+	/// left as `Regular` in most circumstances, but when you want the same power as a `slash`, it
+	/// may be `Force`.
```
`force` (`Fortitude`) now only has possible values `Force` and `Polite`. `Regular` is now `Polite`
```suggestion
	/// If `force` is `Force`, then other fund-locking mechanisms may be disregarded. It should be
	/// left as `Polite` in most circumstances, but when you want the same power as a `slash`, it
	/// may be `Force`.
```

tonyalaribe commented on frame/support/src/traits/tokens/fungible/hold.rs:
```diff
@@ -0,0 +1,393 @@
+		mode: Restriction,
+		force: Fortitude,
+	) -> Result<Self::Balance, DispatchError> {
+		// We must check total-balance requirements if `!force`.
```
```suggestion
		// We must check total-balance requirements if `force` is `Fortitude::Polite`.
```

tonyalaribe commented on frame/support/src/traits/tokens/fungible/hold.rs:
```diff
@@ -0,0 +1,393 @@
+	/// `source` must obey the requirements of `keep_alive`.
+	///
+	/// If `force` is `Force`, then other fund-locking mechanisms may be disregarded. It should be
+	/// left as `Regular` in most circumstances, but when you want the same power as a `slash`, it
```
```suggestion
	/// left as `Polite` in most circumstances, but when you want the same power as a `slash`, it
```

tonyalaribe commented on primitives/runtime/src/lib.rs:
```diff
@@ -619,18 +620,25 @@ pub enum TokenError {
+	/// Account cannot be created for a held balance.
+	CannotCreateHold,
+	/// Withdrawal would cause unwanted loss of account.
+	UnwantedRemoval,
```
This was a little confusing to me when I saw it in the code above. Would something like `UnwantedAccountRemoval` work as a clearer name? It wasn't clear without the docs that the error was refering to the account.

tonyalaribe reviewed (commented):
Just some small comments. But LGTM so far.

KiChjang commented on bin/node/cli/tests/temp_base_path_works.rs:
```diff
@@ -32,6 +32,7 @@ use std::{
 
 pub mod common;
 
+#[ignore]
```
Do we still want to ignore this test?

ggwpez replied:
Was probably randomly failing https://github.com/paritytech/substrate/pull/13505

KiChjang commented on frame/balances/src/lib.rs:
```diff
@@ -232,8 +238,14 @@ pub mod pallet {
 
-		/// Weight information for extrinsics in this pallet.
-		type WeightInfo: WeightInfo;
+		/// The ID type for reserves. Use of reserves is deprecated.
```
Might be better to mention here what the alternatives are for reserves -- I'm guessing it's both hold and freeze?

gavofyork committed:
Update frame/executive/src/lib.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/executive/src/lib.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2487611

gavofyork committed:
Reenable test

gavofyork commented on frame/executive/src/lib.rs:
```diff
@@ -1257,54 +1265,54 @@ mod tests {
-						&weight,
+	/*
+		#[test]
+		fn can_pay_for_tx_fee_on_full_lock() {
```
Reenabled.

gavofyork committed:
Merge branch 'gav-hold-freeze' of github.com:paritytech/substrate into gav-hold-freeze

gavofyork committed:
Update frame/support/src/traits/tokens/fungibles/hold.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungible/hold.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungible/hold.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungible/hold.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/currency.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2488319

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2488321

gavofyork committed:
Update frame/lottery/src/tests.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungible/mod.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungible/regular.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungibles/freeze.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungible/regular.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungibles/hold.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungibles/hold.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Update frame/support/src/traits/tokens/fungibles/hold.rs

Co-authored-by: Anthony Alaribe <anthonyalaribe@gmail.com>

gavofyork committed:
Rename UnwantedRemoval to UnwantedAccountRemoval

gavofyork committed:
Merge branch 'gav-hold-freeze' of github.com:paritytech/substrate into gav-hold-freeze

gavofyork committed:
Docs

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
Formatting

gavofyork commented:
@ruseinov, @KiChjang, @ggwpez thanks for the review/suggestions - they've been addressed. Once you're happy please provide approvals.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2490955

KiChjang commented on primitives/runtime/src/lib.rs:
```diff
@@ -619,18 +620,25 @@ pub enum TokenError {
+	/// Account cannot be created for a held balance.
+	CannotCreateHold,
+	/// Withdrawal would cause unwanted loss of account.
+	UnwantedAccountRemoval,
```
This reads like it's removing an unwanted account, while the meaning here is supposed to be an unwanted removal of an account. How about we reword the variant to the following?

```suggestion
	UnwantedRemovalOfAccount,
```

Or switch `Removal` with `Loss`?

gavofyork committed:
Update frame/balances/src/lib.rs

Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

gavofyork committed:
Update primitives/runtime/src/lib.rs

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514915

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514926

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2514916

tonyalaribe commented on frame/balances/src/lib.rs:
```diff
@@ -603,1601 +467,606 @@ pub mod pallet {
+					upgrade_count.saturating_inc();
+				}
+			}
+			if if Perbill::from_rational(upgrade_count, who.len()) >= Perbill::from_percent(90) { {
```
There was a mistake in the last commit that won't compile
```suggestion
			if Perbill::from_rational(upgrade_count, who.len()) >= Perbill::from_percent(90) {
```

gavofyork commented on frame/assets/src/impl_fungibles.rs:
```diff
@@ -65,69 +77,24 @@ impl<T: Config<I>, I: 'static> fungibles::Inspect<<T as SystemConfig>::AccountId
-	fn set_balance(_: Self::AssetId, _: &T::AccountId, _: Self::Balance) -> DispatchResult {
-		unreachable!("set_balance is not used if other functions are impl'd");
+	fn handle_dust(_: fungibles::Dust<T::AccountId, Self>) {
+		unreachable!("`decrease_balance` and `increase_balance` have non-default impls; nothing else calls this; qed");
```
Better just to reimpl `handle_raw_dust` to do nothing.

gavofyork committed:
handle_raw_dust oes nothing

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2517590

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2517605

gavofyork committed:
Formatting

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2518442

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2518428

gavofyork committed:
Fixes

gavofyork committed:
Grumble

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2518665

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2518651

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2518706

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2518720

gavofyork committed:
Add test

gavofyork committed:
Add test

gavofyork commented on frame/balances/src/impl_currency.rs:
```diff
@@ -0,0 +1,904 @@
+
+				// bail if we need to keep the account alive and this would kill it.
+				let ed = T::ExistentialDeposit::get();
+				let would_be_dead = new_free_account + account.reserved < ed;
```
It is now, and a test added.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2518804

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2518818

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
Tests for reducible_balance

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2521335

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2521384

gavofyork committed:
Fix Salary

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2521444

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2521445

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2521504

gavofyork committed:
Disable broken test

gavofyork committed:
Disable nicely

gavofyork committed:
Fixes

joepetrowski commented on frame/balances/src/lib.rs:
```diff
@@ -477,26 +303,44 @@ pub mod pallet {
+		Restored { who: T::AccountId, amount: T::Balance },
+		/// An account was upgraded.
+		Upgraded { who: T::AccountId },
+		/// Total issuance was increased by `amount`, creating a credit to be balanced.
+		Issue { amount: T::Balance },
+		/// Total issuance was decreased by `amount`, creating a debt to be balanced.
+		Rescind { amount: T::Balance },
```
Keep past participle convention?
```suggestion
		/// Total issuance was increased by `amount`, creating a credit to be balanced.
		Issued { amount: T::Balance },
		/// Total issuance was decreased by `amount`, creating a debt to be balanced.
		Rescinded { amount: T::Balance },
```

gavofyork committed:
Fixes

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2527765

gavofyork committed:
Fixes

gavofyork committed:
Rename some events

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
Fix nomination pools breakage

gavofyork committed:
Add compatibility stub for transfer tx

KiChjang commented on frame/referenda/src/migration.rs:
```diff
@@ -191,7 +191,7 @@ pub mod test {
 	pub fn referendum_status_v0() {
 		// make sure the bytes of the encoded referendum v0 is decodable.
-		let ongoing_encoded = sp_core::Bytes::from_str("0x00000000013001012a000000000000000400000100000000000000010000000000000001000000000000000a00000000000000000000000000000000000100").unwrap();
+		let ongoing_encoded = sp_core::Bytes::from_str("0x00000000012c01012a0000000000000004000100000000000000010000000000000001000000000000000a00000000000000000000000000000000000100").unwrap();
```
What's going on here? Why is there a change in the encoded bytes?

gavofyork committed:
Reinstate a safely compatible version of Balances set_balance

gavofyork committed:
Fixes

gavofyork commented:
bot merge force

ggwpez commented:
bot bench $ pallet dev pallet_balances

gavofyork commented:
bot merge force

gavofyork commented:
bot merge force

ggwpez commented on frame/balances/src/impl_currency.rs:
```diff
@@ -0,0 +1,899 @@
+			 is_new|
+			 -> Result<SignedImbalance<Self::Balance, Self::PositiveImbalance>, DispatchError> {
+				let ed = T::ExistentialDeposit::get();
+				let total = value.saturating_add(account.reserved);
+				// If we're attempting to set an existing account to less than ED, then
+				// bypass the entire operation. It's a no-op if you follow it through, but
+				// since this is an instance where we might account for a negative imbalance
+				// (in the dust cleaner of set_account) before we account for its actual
+				// equal and opposite cause (returned as an Imbalance), then in the
+				// instance that there's no other accounts on the system at all, we might
+				// underflow the issuance and our arithmetic will be off.
+				ensure!(total >= ed || !is_new, Error::<T, I>::ExistentialDeposit);
```
Here the reserved is again taken into account for the ED comparison.

gavofyork committed:
Grumble

kianenigma commented on frame/nis/src/lib.rs:
```diff
@@ -209,7 +213,15 @@ pub mod pallet {
+			+ FunHoldInspect<Self::AccountId>
+			+ FunHoldMutate<Self::AccountId>;
+
+		/// The name for the reserve ID.
```
```suggestion
		/// The identifier of the hold reason.

```

kianenigma commented on frame/nis/src/lib.rs:
```diff
@@ -615,19 +625,21 @@ pub mod pallet {
 
+			let mut queue = Queues::<T>::get(duration);
+			let pos = queue.iter().position(|i| i == &bid).ok_or(Error::<T>::UnknownBid)?;
+			queue.remove(pos);
```
This is a great example of a code that's being rewritten simpler using transactional storage in mind @liamahoran

gavofyork committed:
Update frame/nis/src/lib.rs

Co-authored-by: Kian Paimani <5588131+kianenigma@users.noreply.github.com>

gavofyork commented:
bot merge force

Commit:
Merge branch 'master' of https://github.com/paritytech/substrate into HEAD

Commit:
".git/.scripts/commands/bench/bench.sh" pallet dev pallet_balances

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2540531

kianenigma commented on frame/balances/src/lib.rs:
```diff
@@ -232,8 +238,14 @@ pub mod pallet {
 
-		/// Weight information for extrinsics in this pallet.
-		type WeightInfo: WeightInfo;
+		/// The ID type for reserves. Use of reserves is deprecated in favour of holds.
```
A good follow-up, or in this PR action, would be to actually mark all traits related to locking and reserves as depreacted, sprinkel `#[allow(deprecated)]` where needed, and have clear list of which pallets need to be migrated.

gavofyork replied:
Definitely.

kianenigma replied:
Made an quick issue for now.

kianenigma commented on frame/balances/src/lib.rs:
```diff
@@ -232,8 +238,14 @@ pub mod pallet {
 
-		/// Weight information for extrinsics in this pallet.
-		type WeightInfo: WeightInfo;
+		/// The ID type for reserves. Use of reserves is deprecated in favour of holds.
```
```suggestion
		/// The ID type for reserves. 
		/// 
		/// Use of reserves is deprecated in favour of holds. See `https://github.com/paritytech/substrate/pull/12951/`
```

kianenigma commented on frame/balances/src/types.rs:
```diff
@@ -0,0 +1,157 @@
+	/// Paying system transaction fees.
+	Fee = 0,
+	/// Any reason other than paying system transaction fees.
+	Misc = 1,
```
Why are the indices manually set differently?

gavofyork replied:
Legacy code https://github.com/paritytech/substrate/pull/12951/files/1d47200f500f3799dc940e9ff15b7cf1bdba90ab#diff-7f35e750129ddf17862e61be2ce506a120ac7d4de151633d2c026525130bba7dL629-L638. It hasn't changed in years - just repotted in this PR. Expect the indices there were just a measure of explicitness. It'll be removed along with `Currency` trait once we move everything to `fungible`.

kianenigma commented on frame/balances/src/lib.rs:
```diff
@@ -602,1601 +466,670 @@ pub mod pallet {
+				}
+			}
+			let proportion_upgraded = Perbill::from_rational(upgrade_count, who.len() as u32);
+			if proportion_upgraded >= Perbill::from_percent(90) {
```
Simple follow-up is to write a bot for this.

gavofyork replied:
There's already a script linked in the description courtesy @ggwpez https://github.com/ggwpez/substrate-scripts/blob/master/upgrade-accounts.py

kianenigma commented on frame/balances/src/types.rs:
```diff
@@ -0,0 +1,157 @@
+	pub frozen: Balance,
+	/// Extra information about this account. The MSB is a flag indicating whether the new ref-
+	/// counting logic is in place for this account.
+	pub flags: ExtraFlags,
```
This is a new field right? won't it fail on decode? If so, I was expecting to see a custom `Decode` somewhere, perhaps havn't found it yet.

gavofyork replied:
Actually, it replaces the `fee_frozen` field, and it's exactly the same size. This is how I get around a massive synchronous storage migration.

kianenigma commented on frame/balances/src/lib.rs:
```diff
@@ -602,1601 +466,670 @@ pub mod pallet {
+		///
+		/// The dispatch origin for this call must be `Signed` by the transactor.
+		///
+		/// # <weight>
```
FYI I am pushing for getting rid of `<weight>` comments, ie. https://github.com/paritytech/substrate/pull/13350. You can skip them now or just put them under `## Complexity` if it is useful for the developers to know. If it is only for us, I'd put it under `//` as *implementation note*.

kianenigma commented on frame/balances/src/tests/mod.rs:
```diff
@@ -0,0 +1,299 @@
+	}
+}
+
+thread_local! {
+	pub static SYSTEM_STORAGE: sp_std::cell::RefCell<bool>  = sp_std::cell::RefCell::new(false);
+}
+pub fn use_system() -> bool {
+	SYSTEM_STORAGE.with(|q| *q.borrow())
+}
```
```suggestion
parameter_types! {
	pub static UseSystem: bool = false;
}

// if UseSystem::get() ...
// UseSystem::set(true);
```

I've tried to communicate this new syntax to you a few times in reviews 😅 not sure if it was not seen, or you prefer to old school syntax?

kianenigma replied:
under the hood they it generates the same `thread_local` code.

gavofyork replied:
Noted.

kianenigma commented on frame/election-provider-multi-phase/src/mock.rs:
```diff
@@ -254,6 +254,10 @@ impl pallet_balances::Config for Runtime {
 	type MaxReserves = ();
 	type ReserveIdentifier = [u8; 8];
 	type WeightInfo = ();
+	type FreezeIdentifier = ();
```
Have you seen https://github.com/paritytech/substrate/pull/13454, and any opinions on it? It would have made the diff for this PR significantly smaller, assuming that you would think it is safe to use these defaults everywhere.

gavofyork replied:
How far off merge is it?

kianenigma replied:
I'd guess a week or two, but @sam0x17 can verify. 

The base PR can probably be merged soon, without using it extensively.

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork committed:
disable flakey tests

gavofyork committed:
Update frame/balances/src/lib.rs

Co-authored-by: Kian Paimani <5588131+kianenigma@users.noreply.github.com>

gavofyork committed:
Grumbles

gavofyork committed:
Merge branch 'gav-hold-freeze' of github.com:paritytech/substrate into gav-hold-freeze

gavofyork committed:
Grumble

kianenigma reviewed (approved):
Looking forward to start migrating pallets one by one as well ie #13633

gavofyork committed:
Merge branch 'master' into gav-hold-freeze

gavofyork committed:
Merge remote-tracking branch 'origin/master' into gav-hold-freeze

gavofyork commented:
bot merge force

gavofyork commented:
bot merge force

xlc commented on frame/balances/src/lib.rs:
```diff
@@ -602,1601 +468,646 @@ pub mod pallet {
+					// Gah!! We have a non-zero reserve balance but no provider refs :(
+					// This shouldn't practically happen, but we need a failsafe anyway: let's give
+					// them enough for an ED.
+					a.free = a.free.min(T::ExistentialDeposit::get());
```
`min` or `max`?
also does this just mint ED without update total issuance?

gavofyork replied:
Good catch on the `min`: https://github.com/paritytech/substrate/pull/13741

It doesn't update total issuance, but as per the comment: this is an emergency measure shich should never be called in practice.

xlc commented on frame/nis/src/lib.rs:
```diff
@@ -473,6 +480,9 @@ pub mod pallet {
 		AlreadyPrivate,
+		Release1,
+		Release2,
+		Tah,
```
??

gavofyork replied:
Removed in #13741

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-41-v0-9-42/2828/1

shawntabrizi commented:
> within 3 months at which point the transfer alias will be removed.

Looks like it should be removed by now

liamaharon commented:
> > within 3 months at which point the transfer alias will be removed.
> 
> Looks like it should be removed by now

cc @juangirini

juangirini commented:
> > within 3 months at which point the transfer alias will be removed.
> 
> Looks like it should be removed by now

https://github.com/paritytech/substrate/issues/14652
cc @liamaharon
